zebROS_ws/src/candle_controller/src/candle_controller.cpp

```diff
@@ -0,0 +1,255 @@
+        }
+        case 8: {
+            return CANdleAnimationType::TwinkleOff;
+        }
+        default: {
+            ROS_ERROR_STREAM("Invalid int to convert to CANdleAnimationType!");
+        }
+    }
+}
+CANdleAnimationClass getAnimationClass(CANdleAnimationType type) {
+    switch (type) {
+        case CANdleAnimationType::ColourFlow: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        case CANdleAnimationType::Fire: {
+            return CANdleAnimationClass::BaseStandard;
+        }
+        case CANdleAnimationType::Larson: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        case CANdleAnimationType::Rainbow: {
+            return CANdleAnimationClass::BaseStandard;
+        }
+        case CANdleAnimationType::RGBFade: {
+            return CANdleAnimationClass::BaseStandard;
+        }
+        case CANdleAnimationType::SingleFade: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        case CANdleAnimationType::Strobe: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        case CANdleAnimationType::Twinkle: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        case CANdleAnimationType::TwinkleOff: {
+            return CANdleAnimationClass::BaseTwo;
+        }
+        default: {
+            ROS_ERROR_STREAM("Faild to match CANdleAnimationType to get CANdleAnimationClass!");
+        }
+    }
+}
+
+namespace candle_controller {
+class CANdleController : public controller_interface::Controller<CANdleCommandInterface> {
+public:
+    CANdleController() {}
+
+    bool init(
+        CANdleCommandInterface* candle_command_interface,
+        ros::NodeHandle& /*root_nh*/,
+        ros::NodeHandle& controller_nh
+    ) override {
+        std::string candle_name;
+        if (!controller_nh.getParam("name", candle_name)) {
+            ROS_ERROR("Cannot initialize candle! Failed to read the 'name' field.==========");
+            return false;
+        }
+
+        this->candle_handle = candle_command_interface->getHandle(candle_name);
+
+        this->colour_service = controller_nh.advertiseService("colour", &CANdleController::colourCallback, this);
+        this->brightness_service = controller_nh.advertiseService("brightness", &CANdleController::brightnessCallback, this);
+        this->animation_service = controller_nh.advertiseService("animation", &CANdleController::animationCallback, this);
+        return true;
+    }
+
+    void starting(const ros::Time&) override {}
+
+    void update(const ros::Time&, const ros::Duration&) override {
+        const LEDs leds = *(this->led_buffer.readFromRT());
+        const Animation animation = *(this->animation_buffer.readFromRT());
+        const Brightness brightness = *(this->brightness_buffer.readFromRT());
+
+        // Brightness isn't exclusive to colours/animations, so it gets special treatment
+        if (brightness.time > this->last_write) {
```

Comment:
It might be OK to rewrite brightness each time through as long as the lower level command code does an early out if new_brightness == current_brightness.  That would simplify the controller code a good bit.

Reply:
I think the main reason I had the if statement was for the debug message. Once I'm confident this all works, we can remove that and just write brightness each time.

```c++
if (brightness.time > this->last_write) {
    ROS_INFO_STREAM("Writing new brightness to CANdle");
    this->candle_handle->setBrightness(brightness.brightness);
}
```